onadata/apps/logger/migrations/0030_backfill_lost_monthly_counters.py

```diff
@@ -0,0 +1,74 @@
+from django.db import migrations
+from django.db.migrations.recorder import MigrationRecorder
+from django.db.models import Sum
+from django.db.models import Value, F, DateField
+from django.db.models.functions import Cast, Concat
+from django.db.models.functions import ExtractYear, ExtractMonth
+from django.utils.timezone import now
+
+
+def populate_missing_monthly_counters(apps, schema_editor):
+
+    DailyXFormSubmissionCounter = apps.get_model('logger', 'DailyXFormSubmissionCounter')  # noqa
+    MonthlyXFormSubmissionCounter = apps.get_model('logger', 'MonthlyXFormSubmissionCounter')  # noqa
+
+    if not DailyXFormSubmissionCounter.objects.all().exists():
+        return
+
+    previous_migration = MigrationRecorder.Migration.objects.filter(
+        app='logger', name='0029_populate_daily_xform_counters_for_year'
+    ).first()
+
+    # Delete monthly counters in the range if any (to avoid conflicts in bulk_create below)
+    MonthlyXFormSubmissionCounter.objects.annotate(
+        date=Cast(
+            Concat(
+                F('year'), Value('-'), F('month'), Value('-'), 1
+            ),
+            DateField(),
+        )
+    ).filter(date__gte=previous_migration.applied.date().replace(day=1)).delete()
+
+    records = (
+        DailyXFormSubmissionCounter.objects.filter(
+            date__range=[
+                previous_migration.applied.date().replace(day=1),
+                now().date()
+            ]
+        )
+        .annotate(year=ExtractYear('date'), month=ExtractMonth('date'))
+        .values('month', 'year')
+        .annotate(total=Sum('counter'))
+        .values('user_id', 'xform_id', 'month', 'year', 'total')
+    ).order_by('year', 'month', 'user_id')
```

Comment:
I think this is a little bit confusing to read and that the first call to `values()` doesn't really do anything. The SQL ends up being generated with a single `GROUP BY` that reflects the second call to `values()`:

```sql
SELECT "logger_dailyxformsubmissioncounter"."user_id",
       "logger_dailyxformsubmissioncounter"."xform_id",
       EXTRACT('year' FROM "logger_dailyxformsubmissioncounter"."date") AS "year",
       EXTRACT('month' FROM "logger_dailyxformsubmissioncounter"."date") AS "month",
       SUM("logger_dailyxformsubmissioncounter"."counter") AS "total"
  FROM "logger_dailyxformsubmissioncounter"
 WHERE "logger_dailyxformsubmissioncounter"."date" BETWEEN '2023-10-01'::date AND '2023-10-09'::date
 GROUP BY EXTRACT('year' FROM "logger_dailyxformsubmissioncounter"."date"),
          EXTRACT('month' FROM "logger_dailyxformsubmissioncounter"."date"),
          "logger_dailyxformsubmissioncounter"."user_id",
          "logger_dailyxformsubmissioncounter"."xform_id"
 ORDER BY "year" ASC,
          "month" ASC,
          "logger_dailyxformsubmissioncounter"."user_id" ASC
```

A simplification to this produces the same SQL with a caveat (see below):
```python
records = (
    DailyXFormSubmissionCounter.objects.filter(
        date__range=[
            previous_migration.applied.date().replace(day=1),
            now().date()
        ]
    )
    .annotate(year=ExtractYear('date'), month=ExtractMonth('date'))
    .values('user', 'xform', 'month', 'year')
    .annotate(total=Sum('counter'))
).order_by('year', 'month', 'user')
```

The caveat is that the ORM changes the order of the columns in the `GROUP BY`, which [I believe](https://stackoverflow.com/questions/3064677/does-the-order-of-columns-matter-in-a-group-by-clause) does not matter.
```diff
--- before.sql  2023-10-09 03:21:16.638677449 -0400
+++ after.sql   2023-10-09 03:21:23.255101811 -0400
@@ -1,15 +1,15 @@
 SELECT "logger_dailyxformsubmissioncounter"."user_id",
        "logger_dailyxformsubmissioncounter"."xform_id",
        EXTRACT('year' FROM "logger_dailyxformsubmissioncounter"."date") AS "year",
        EXTRACT('month' FROM "logger_dailyxformsubmissioncounter"."date") AS "month",
        SUM("logger_dailyxformsubmissioncounter"."counter") AS "total"
   FROM "logger_dailyxformsubmissioncounter"
  WHERE "logger_dailyxformsubmissioncounter"."date" BETWEEN '2023-10-01'::date AND '2023-10-09'::date
- GROUP BY EXTRACT('year' FROM "logger_dailyxformsubmissioncounter"."date"),
-          EXTRACT('month' FROM "logger_dailyxformsubmissioncounter"."date"),
-          "logger_dailyxformsubmissioncounter"."user_id",
-          "logger_dailyxformsubmissioncounter"."xform_id"
+ GROUP BY "logger_dailyxformsubmissioncounter"."user_id",
+          "logger_dailyxformsubmissioncounter"."xform_id",
+          EXTRACT('year' FROM "logger_dailyxformsubmissioncounter"."date"),
+          EXTRACT('month' FROM "logger_dailyxformsubmissioncounter"."date")
  ORDER BY "year" ASC,
           "month" ASC,
           "logger_dailyxformsubmissioncounter"."user_id" ASC
```

I removed the `_id` suffix in the ORM queries where it's unnecessary for Django, but I don't mind if you prefer to leave it in.

Reply:
I kept `_id` because that's what the dict returns anyway. It's more obvious to me.